### PR TITLE
Stories/2492/issues/53 mechanism to control value range of nodes

### PR DESCRIPTION
**Pull request by suneetnangia**
## Purpose
Added feature to allow random and sequential node value
* Double Type Node: Random within range (with all ranges e.g. -5.0 to +10.2)
* Double Type Node: Sequential increment with configurable step size (positive range e.g. 0 to +10.2 )
* UInt Type Node: Random within range (with all ranges e.g. 0 to +10)
* UInt Type Node: Sequential increment with configurable step size (positive range e.g. 0 to +10)

## Does this introduce a breaking change?
```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone [repo-address]
cd [repo-name]
git checkout [branch-name]
npm install
```

* Test the code

```
```
## Other Information
Still working on adding negative range for double type in sequential increment case e.g. -5.0 to +5.0 with recycling.

### CODE REVIEW

File: src/Program.cs [hunk 1]
```diff
@@ -544,18 +576,9 @@ private static async Task ConsoleServerAsync(CancellationToken cancellationToken
...
-            Console.CancelKeyPress += (_, eArgs) => {
-                cancellationTokenSource.Cancel();
-                eArgs.Cancel = true;
-            };
-            while (!cancellationTokenSource.Token.WaitHandle.WaitOne(1000))
-            {
-            }
+            quitEvent.WaitOne(Timeout.Infinite);
             PlcSimulation.Stop();
-            PlcServer.Stop();
```
Comment (algattik) [inline src/Program.cs:558]: why this change (and other changes in this class)? Merge issue?

Comment (suneetnangia) [reply src/Program.cs]: Resolved, it was merge issue.

File: README.md [hunk 2]
```diff
@@ -96,9 +96,17 @@ A number of changing nodes can be simulated with the following options. The node
 - sn: Number of slow nodes (default 1)
 - sr: Rate in seconds at which to change the slow nodes (uint, default every 10 s)
 - st: Data type for slow nodes (UInt|Double|Bool|UIntArray, case insensitive)
+- stl: lower bound of data type of slow nodes (UInt|Double types only, defaults to minimium value of the type in C#)
```
Comment (peterbryntesson) [inline README.md:99]: Four Double, does that default to 0.0 or Double.MinValue?

File: src/PlcNodeManager.cs [hunk 3]
```diff
@@ -434,23 +498,23 @@ private BaseDataVariableState[] CreateBaseLoadNodes(FolderState dataFolder, stri
...
+        private static (NodeId dataType, int valueRank, object defaultValue, object stepSize, object minValue, object maxValue) GetNodeType(NodeType nodeType, string stepSize, string minValue, string maxValue)
         {
             return nodeType switch
             {
-                NodeType.Bool => (new NodeId((uint)BuiltInType.Boolean), ValueRanks.Scalar, true),
-                NodeType.Double => (new NodeId((uint)BuiltInType.Double), ValueRanks.Scalar, (double)0.0),
-                NodeType.UIntArray => (new NodeId((uint)BuiltInType.UInt32), ValueRanks.OneDimension, new uint[32]),
-                _ => (new NodeId((uint)BuiltInType.UInt32), ValueRanks.Scalar, (uint)0),
+                NodeType.Bool => (new NodeId((uint)BuiltInType.Boolean), ValueRanks.Scalar, true, null, null, null),
+                NodeType.Double => (new NodeId((uint)BuiltInType.Double), ValueRanks.Scalar, (double)0.0, double.Parse(stepSize), minValue == null ? double.MinValue : double.Parse(minValue), maxValue == null ? double.MaxValue : double.Parse(maxValue)),
```
Comment (peterbryntesson) [inline src/PlcNodeManager.cs:515]: double.MinValue is negative. Maybe you should use 0.0 here instead.

File: src/PlcNodeManager.cs [hunk 4]
```diff
@@ -502,56 +566,76 @@ private void SetHeaterOffMethodProperties(ref MethodState method)
...
+            baseDataVariableState.WriteMask = AttributeWriteMask.DisplayName | AttributeWriteMask.Description;
+            baseDataVariableState.UserWriteMask = AttributeWriteMask.DisplayName | AttributeWriteMask.Description;
+            baseDataVariableState.DataType = dataType;
+            baseDataVariableState.ValueRank = valueRank;
+            baseDataVariableState.AccessLevel = accessLevel;
+            baseDataVariableState.UserAccessLevel = accessLevel;
+            baseDataVariableState.Historizing = false;
+            baseDataVariableState.Value = defaultValue ?? Opc.Ua.TypeInfo.GetDefaultValue(dataType, valueRank, Server.TypeTree);
+            baseDataVariableState.StatusCode = StatusCodes.Good;
+            baseDataVariableState.Timestamp = DateTime.UtcNow;
```
Comment (peterbryntesson) [inline src/PlcNodeManager.cs:624]: Here you should use TimeService.UtcNow() (see removed code above)

**Review by peterbryntesson — CHANGES_REQUESTED** [review]
LGTM in general - see my comments

File: src/PlcNodeManager.cs (hunk 3 above)
Comment (suneetnangia) [reply src/PlcNodeManager.cs:515]: I think it's a good idea, will help reducing upfront parameters, will update.
